aten/src/ATen/native/cuda/Reduce.cuh

```diff
@@ -373,28 +390,86 @@ struct ReduceOp {
     }
   }
 
-  C10_DEVICE arg_t thread_reduce(const scalar_t* data) const {
+  C10_DEVICE arg_t header_reduce(arg_t value, const scalar_t* &data, index_t &end) const {
+    const int align_bytes = alignof(at::native::memory::aligned_vector<scalar_t, 4>);
+    const int shift = ((uint64_t)data) % align_bytes / sizeof(scalar_t);
+    if (shift > 0) {
+      data -= shift;
+      end += shift;
+      if(threadIdx.x >= shift && config.should_reduce_tail()){
+        value = ops.reduce(value, data[threadIdx.x], threadIdx.x - shift);
+      }
+      end -= blockDim.x;
+      data += blockDim.x;
+    }
+    return value;
+  }
+
+  C10_DEVICE arg_t vectorized_thread_reduce(arg_t value, const scalar_t* data, index_t end) const {
+    using load_t = at::native::memory::aligned_vector<scalar_t, 4>;
+
+    index_t idx = config.input_idx();
+    const index_t stride = config.step_input;
+
+    // Multiple accumulators to remove dependency between unrolled loops.
+    arg_t value_list[4];
+    value_list[0] = value;
+    #pragma unroll
+    for (int i = 1; i < 4; i++) {
+      value_list[i] = ident;
+    }
+
+    scalar_t values[4];
+    load_t *values_vector = reinterpret_cast<load_t*>(values);
+
+    while (idx * 4 + 3 < end) {
+      *values_vector = reinterpret_cast<const load_t*>(data)[idx];
+      #pragma unroll
+      for (index_t i = 0; i < 4; i++) {
+        value_list[i] = ops.reduce(value_list[i], values[i], idx * 4 + i);
+      }
+      idx += stride;
+    }
+
+    // tail
+    index_t tail_start = end - end % 4;
+    if (config.should_reduce_tail()) {
+      int idx = tail_start + threadIdx.x;
+      if (idx < end) {
+        value_list[0] = ops.reduce(value_list[0], data[idx], idx);
+      }
+    }
+
+    // combine accumulators
+    #pragma unroll
+    for (int i = 1; i < 4; i++) {
+      value_list[0] = ops.combine(value_list[0], value_list[i]);
+    }
+    return value_list[0];
+  }
+
+  C10_DEVICE arg_t thread_reduce(arg_t value, const scalar_t* data, index_t end) const {
```

Comment:
why did thread_reduce need to change (e.g. get an `end` argument? It's always called with `end=config.num_inputs` as far as I can see.

Reply:
You are right, I was originally putting `vectorized_thread_reduce` inside `thread_reduce`, therefore the change. But then I moved it out, and forget to change it back.

Reply:
fixed